Contractual Disappointment in C++ (#51)
johnmcfarlane/papers: c++/error-handling.md

a _provider_ about the run-time behaviour of some or all of a program.
_Violation_ of a contract is what leads to disappointment.

For the sake of simplicity and focus,
I will concentrate on the obligations imposed on the user by the contract.
I believe that extending the principles laid out here
to obligations imposed on the provider is relatively straight-forward.

## What Are the Parameters of a Contract?

The following set of parameters -- expressed as questions -- are helpful in
understanding the correct error-handling strategy for a contract.

* Where, and in what form, is the agreement expressed?
* Who is the provider of the contract?
* Who is the user of the contract?
* What are the consequences of user contract violation?

## What Contracts Are There?

In a C++ program, some contracts that matter are as follows.

### End User Contract

The ultimate contract that a program must fulfil is to its end user.
All other contracts below are in support of this fulfilment.

* agreement: program documentation, possibly including a `--help` option
  or a 'man' page
* provider: the program author(s)
* user: the program user

> * violation: user input is sanitized and ill-formed input is handled through

Author: Looks like a definition of `violation` has been skipped, you jump straight into constraints that need to be fulfilled - the reader has to catch up to your thinking somewhat
(from Alicja)

Author: This is now "violation by user" which I think is half the battle. Will get the reviewer to re-read this section later.

>   normal control flow (including exception handling)

Reviewer: This describes the case when program user fails to comply with their obligations (like a precondition). It does not mention the case where the program author fails to deliver their part of the job (like a postcondition).

Author: As mentioned [here](https://github.com/johnmcfarlane/papers/pull/51/files#diff-046c70b4ff393727ea22013b061f963e90339ce00073ea06899f02cd54c46ceaR38):

> For the sake of simplicity and focus, I will concentrate on the obligations imposed on the user by the contract.

I'm not sure what to do to make this clearer and easier to dip into. Possible options:

1. Reword/repeat the limitations to this section.
1. Include both _user_ and _provider_ violations under _violation_.
1. Rename _violation_ to _user violation_.
1. Add a _provider violation_ parameter to "What Are the Parameters of a Contract?" and explain this is omitted.
1. Add a _provider violation_ parameter to "What Are the Parameters of a Contract?" and then fill it out.

One problem is that I am concentrating on what is needed to help the user get things right and I want to keep the amount of detail low. So probably 3) is worth considering.

Another problem is that this document talks about what a program author should do when they encounter different types of disappointment. I generally don't think it's good for user code to be testing provider violations, e.g.:

```c++
std::vector v;
ASSERT(v.empty());  // user shouldn't do this
```

OTOH, maybe spelling this out is a good idea!

Reviewer: To focus only on precondition-like things seems fine to me.  But the next section, "Test User Contract" seems to be talking about something closer to postconditions. And this is where I get confused (about the scope of this paper).

Author: You're right. It's the most hard-to-understand section and I got a lot of it wrong. I've rewritten it. Try to forget you read it yet and see if it makes any sense now.

User input might include command-line parameters, UI interaction, data files.
Input identified as ill-formed must be rejected by the program.

> Where possible rejection should be accompanied by meaningful feedback

Author: ```suggestion
Where possible, rejection should be accompanied by meaningful feedback
```
(from Alicja)

which helps the user correct the input so as to increase the chance of future success.
For example, on a typical terminal-based system,
meaningful feedback might take the form of a diagnostic emitted on the error stream,
followed by exit with non-zero status.

Input that is not rejected must be incapable of causing violations
of any contracts mentioned below.
In particular, contract violations could result in security vulnerabilities.
For programs with exposure to malicious agents, this is critical.
For safety-critical applications, this is important.

> Fuzz testing is a method which helps identify flaws in input sanitization.

Author: ```suggestion 
To identify flaws in input sanitization use fuzz testing.
```
I think the text would flow better like this (Alicja)

Of these, a subset can be expressed formally within the program
using assertions (including API preconditions and postconditions).

These expressions can be evaluated at run-time but typically incur a cost
which violates the zero-overhead principle in the case of a bug-free program.

Broadly, there are three possible strategies that can be adopted
at the point where an assertion would evaluate to `false`:

### Log-And-Continue Enforcement Strategy

The assert statement could emit a run-time diagnostic regarding the contract violation
and then continue past the assertion.

This is unwise: the program is known to exhibit undefined behaviour at this
point. If the provider were able to identify a reason why a violation might have
occurred they would have identified a bug. If the provider had identified a bug,
they should have fixed the bug. If the provider had fixed the bug,
they would no longer be able to identify a reason why a violation might have occurred.
Thus any violation is beyond the comprehension of the provider and "all bets are
off" regarding the behaviour of the program. This is essentially what UB means.

> Nevertheless, under certain constraints

Reviewer: Looks like unfinished sentence.

Author: Finished

### Trap Enforcement Strategy

The assert statement can halt the program.

This has the disadvantage that the program completely fails to perform its task.
But the advantage is that no further incorrect behaviour
will be exhibited by the program.

### Prevention Enforcement Strategy

> Assuming bugs are eliminated, no assert expression will ever be false.

Reviewer: > Assuming bugs are eliminated ...

I think, the core of the of the disagreement (whether to allow or disallow contract-based optimizations) lies in whether one can make this assumption. No amount of runtime testing (as opposed to a static correctness proof) can give a certainty that the program is bug-free. Also, even if we were able to run our program with all possible combinations of input and obtain the expected results, it is not a guarantee that there was no Abstract Machine Violation. The same binary that is later run on a minimally different architecture may expose different behavior.

I agree with your statement that only "bugs cause undefined behaviour", but there might be a HyrumBug situation at the abstract machine level: we have UB and the program works fine. But if we add contract-based optimizations to the mix, we break it.

Of course, an individual programmer (like a video game developer) may say, "I am willing to take the risk", and I am not arguing with this choice. But I am far from recommending it to the wider audience.

Author: I disagree with none of this and it's a central concern to all involved. However, one already cannot make this assumption in C++20. We already enable our optimisers, our code already has UB, and we survive.

Software is complex enough that it's no longer a matter of "100% bug free". It's a question of "how many 9's"?

Contracts +  UB + sanitizers + TDD gets you enough 9's. Log+continue loses 9's. Fuzz testing gets you a bunch more 9's.

> I agree with your statement that only "bugs cause undefined behaviour", but there might be a HyrumBug situation at the abstract machine level: we have UB and the program works fine. But if we add contract-based optimizations to the mix, we break it.

Yup. You don't do this unless you're foolhardy or you're serious about the 9's. I don't ever want to force anyone to do this. 

The main message I'd like to get across in this document is that while

1. `UB = faster + risk` when you enable the optimiser,
2. `UB = slower + certainty` when you enable the sanitizer.

If you only do 1) and not 2), yes the situation is worse with contract optimisations (which are optional). But if you do 2) and then 1), you should get safer, faster code.

This means that assert statements behave as if they are not there.
This in turn means that they can be empty or used as hints to the optimiser.

There is concern ([P2064](https://wg21.link/p2064r0)) that assuming no violations
is somehow incorrect. This is founded on the acceptance of bugs in production code
and the assumption that optimisers *cause* undefined behaviour. They do not. Bugs
cause undefined behaviour. If developers wish to release code with bugs to production,
they are welcome to disable such assumptions. But they are invited, instead, to follow
practices described as part of the Test User Contract which should help identify
such bugs early on.

***The author strongly encourages the use of assumptions in the implementation of
assertions in order to unify the set of Dynamically-Enforceable Contracts and to
simplify the implementation of the Test User Contract.***
In practice, this entails marking violated assertions as unreachable

```c++
#define ASSERT(cond) ((cond) ? static_cast<void>(0) : __builtin_unreachable())
```

which is semantically sound, given violations already must never occur in correctly-written
C++ programs.

> This strategy provides the maximum leeway to support enforcement profiles (below)

Reviewer: I do not think that implementing contracts annotations with `__builtin_unreachable()` is necessary for enabling profilers.
The provision that upon failed runtime check a violation handler is invoked is enough to implement a feature in the UB-sanitizer that detects not only Abstract Machine Violations but also user contrct violations.

Constant expressions can also be handled directly: we could say that if the evaluation of a core constant expression invokes the violation handler, the program is ill formed.

So the only benefit that `__builtin_unreachable()` would give is the potential to implement tricky optimizations.
Personally, I am not against such optimizations, but I do not see a need to actively support them in the language. If such optimizations are really practical, compilers will implement them anyway, even if this is not standard-compliant. But I think it is standard-compliant if you install a violation handler with `__builtin_unreachable()`.

Author: > I do not think that implementing contracts annotations with `__builtin_unreachable()` is necessary for enabling profilers.

Do you mean a program profiling tool? I'm referring to *Enforcement Profiles* as listed below. These are different common use cases with which I hope to cover most users' needs. Perhaps I should choose a different term for these.

> The provision that upon failed runtime check a violation handler is invoked is enough to implement a feature in the UB-sanitizer that detects not only Abstract Machine Violations but also user contrct violations.

A violation handler isn't "enough", but excessive. C++20 using an `ASSERT` macro already does everything you describe. 

Are you saying that a handler (say, the one in P2339R0) would be invoked on violation of an ISO C++ Standard contract violation, e.g. divide-by-zero or null terminator dereference? I wasn't aware of such a suggestion. A C++20 sanitizer will trap such a violation. I wasn't aware that a P2339R0 violation handler would.

IOW, future proposals are a step backwards in terms of usability: the landscape will be bifurcated between handler-based violation and UB sanitizers. Inserting `__builtin_unreachable()` into a handler then becomes extra effort the user needs to go to -- just to get back to where she was in C++20.

> Constant expressions can also be handled directly: we could say that if the evaluation of a core constant expression invokes the violation handler, the program is ill formed.

Again, the C++20 user already enjoys compile-time diagnostics. Handler proposals offers the user something they already have -- at additional effort.

> So the only benefit that `__builtin_unreachable()` would give is the potential to implement tricky optimizations.
Personally, I am not against such optimizations, but I do not see a need to actively support them in the language. If such optimizations are really practical, compilers will implement them anyway, even if this is not standard-compliant. 

We should not be explicitly trying to optimise, and `__builtin_unreachable()` does not do that. But the toolchain user should continue to agree that "-O1 means that bugs are fixed, including all Dynamically-Enforceable Contracts" and that Dynamically-Enforceable API Contracts are a new category of these.

> But I think it is standard-compliant if you install a violation handler with `__builtin_unreachable()`.

I'm not opposed to such a handler, but I don't think that they add anything to the status quo. I see them as something needed to satisfy users who wish to log [BizBugs](http://open-std.org/JTC1/SC22/WG21/docs/papers/2021/p2358r0.pdf#7) without subjecting them to the rigour of a *true* contract violation.

---

I'll try to incorporate some of this into the discussion section somehow, which is the least baked section so far. I've got to think about the points I'm trying to make and the audience I'm trying to reach. Thanks for looking at this!

Reviewer: My apologies. I think I typed this too late in the night, and I introduced an unnecessary confusion. You wrote, "Enforcement Profiles", I read it as "UB-sanitizers" and wrote back "profilers". The remainder of my comment had better be ignored.

Reviewer: Let me take another attempt at communicating my thoughts.

First. It is not clear why it is called "Prevention Enforcement Strategy", how is using `__builtin_unreachable()` more preventive than the other listed strategies?

Second. I think that [P2064](https://wg21.link/p2064r0) doesn't do a good job at communicating legitimate concerns about turning contract violations into Abstract Machine Violations. I will try to present my arguments in a subsequent comment.

Third. I read this article as saying that the semantics of `__builtin_unreachable()` gives implementations enough power to implement all the Enforcement Profiles that you list. (I agree.) Now, in the defense of the current proposals, such as [P2182](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p2182r1.html), they use a more complex specification in order to (1) allow the same (and more) Enforcement Profiles but (2) make sure that optimizations based on contract annotations is not one of them. They use terms like "install a violation handler (in an implementation-defined way)", but this does not mean that someone will have to define a function with a concrete signature.  For instance, a GCC could give an option to compile with UB-sanitizer, and this UB-sanitizer reports both Abstact Machine Violations and business contract violations (expressed with contract annotations). This happens with one compiler switch `-fsanitize=undefined`. Such compiler feature would be compliant with [P2182](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p2182r1.html): the compiler switch would be this "implementation-defined way" of installing the handler. IOW, the burden is on the people who read the standardese, not on the programmers who use the compiler interface.

Author: > First. It is not clear why it is called "Prevention Enforcement Strategy", how is using __builtin_unreachable() more preventive than the other listed strategies?

Anything that aids testing is preventative and trapping is the best approach here so `__builtin_unreachable()` in combination with a sanitizer is as good as anything else. OTOH, log-and-continue is likely to be less preventative. It's like the difference between errors and warnings: logging messages can be ignored.

> IOW, the burden is on the people who read the standardese, not on the programmers who use the compiler interface.

It's not 100% clear to me. Does P2064 propose wording to support custom handlers in a portable way or does it simply leave the door open for implementation-defined handlers?

Reviewer: > Anything that aids testing is preventative

Agreed so far.

> and trapping is the best approach here

This is confusing. My comment is regarding section "Prevention Enforcement Strategy". You argue that trapping is the best enforcement strategy. But there is a previous section "Trap Enforcement Strategy". Your comment explains the previous section well, but does not explain the current one ("Enforcement"). 

> so `__builtin_unreachable()` in combination with a sanitizer is as good as anything else.

If you used `__builtin_trap()` for assisting the sanitizer, I would be on the same page. But you use `__builtin_unreachable()`, which I believe does not have the trapping semantics. My experience (with one of the UB-sanitizers, and I think clang static analyzer as well) is that even these tools take the things you declare with `__builtin_unreachable()` for granted and do not try to sanitize/detect if what you assert with it is incorrect.

> OTOH, log-and-continue is likely to be less preventative.

I agree. But I thing this is beside the point of disagreement. My point of confusion is: what do you consider to be the difference between "Trap Enforcement Strategy" and "Prevention Enforcement Strategy"? Why is `__builtin_unreachable()` more preventive than `__builtin_trap()`?

> > IOW, the burden is on the people who read the standardese, not on the programmers who use the compiler interface.

> It's not 100% clear to me. Does P2064 propose wording to support custom handlers in a portable way or does it simply leave the door open for implementation-defined handlers?

P2064 is "Assumptions". The recent papers that propose solutions are

* [P2182R1](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p2182r1.html) "Contract Support: Defining the Minimum Viable Feature Set",
* [P2388R0](https://isocpp.org/files/papers/P2388R0.html) "Abort-only contract support" (submitted only Today).

They both allow (but not require) the implementations to provide a way to install a custom violation handler. How the interface for installing such a handler looks like is not defined: it is left for implementations to decide. I can say it with confidence, because I co-authored both. But it is possible that this intent is not reflected clearly enough in these papers.

(The difference between the two papers is that the later does not allow throwing or continuing violation handlers.)

Author: Too many thread!

> If you used __builtin_trap() for assisting the sanitizer, I would be on the same page. But you use __builtin_unreachable(), which I believe does not have the trapping semantics. My experience (with one of the UB-sanitizers, and I think clang static analyzer as well) is that even these tools take the things you declare with __builtin_unreachable() for granted and do not try to sanitize/detect if what you assert with it is incorrect.

The reason Test User Contract is so difficult to convey is because it changes the rules. Trapping is desired in one case (Test User Contract) but not necessarily in another (Release, End User Contract). Assuming sanitizers are enabled in test builds, `__builtin_unreachable()` does a different thing in both cases but it also does the *right* thing in both cases.

I've rewritten the Test Case Contract section. I hope it's closer to helpful.

> what do you consider to be the difference between "Trap Enforcement Strategy" and "Prevention Enforcement Strategy"?

"Trap Enforcement Strategy":
* In a production build:
  * program **traps on** C++ API Contract violation
  * program assumes no ISO C++ Standard violation
* In a test build (w. all the sanitizer goodness):
  * program traps on C++ API Contract violation
  * program traps on ISO C++ Standard violation

Prevention Enforcement Strategy:
* In a production build:
  * program **assumes no** C++ API Contract violation
  * program assumes no ISO C++ Standard violation
* In a test build (w. all the sanitizer goodness):
  * program traps on C++ API Contract violation
  * program traps on ISO C++ Standard violation

A program author already **must** sanitize before they optimise because the program author needs to prevent UB. That's a C++20 fact of life. With the Prevention Enforcement Strategy, the same rule applies to C++ API Contracts. They must be ruled out.

> Why is __builtin_unreachable() more preventive than __builtin_trap()?

If you aim to prevent bugs before production, it's moot that bugs trap because no bugs -> no trapping. Therefore `__builtin_unreachable()` is the right choice because it goes away in a non-sanitizer build and produces more efficient code.

But if you didn't fix your bugs before production, you really need to keep the bug-handling code there somehow. (Depends on the Enforcement Profile.)

> I can say it with confidence, because I co-authored both. But it is possible that this intent is not reflected clearly enough in these papers.

I am largely on-board with those papers but I just don't think my view of software development is familiar beyond my ~victims~ colleagues. ;P I can believe it's alien to Bloomberg devs, given John's and Joshua's very extensive set of macros and Microsoft devs, given how many bugs the developers in p2064r0 let into production. I'm not saying my code is anywhere near perfect but I find it hard to believe those devs used sanitizers. I don't think that's viable any more.

is somehow incorrect. This is founded on the acceptance of bugs in production code
and the assumption that optimisers *cause* undefined behaviour. They do not. Bugs
cause undefined behaviour. If developers wish to release code with bugs to production,
they are welcome to disable such assumptions. But they are invited, instead, to follow
practices described as part of the Test User Contract which should help identify
such bugs early on.

***The author strongly encourages the use of assumptions in the implementation of
assertions in order to unify the set of Dynamically-Enforceable Contracts and to
simplify the implementation of the Test User Contract.***
In practice, this entails marking violated assertions as unreachable

```c++
#define ASSERT(cond) ((cond) ? static_cast<void>(0) : __builtin_unreachable())
```

which is semantically sound, given violations already must never occur in correctly-written
C++ programs.

This strategy provides the maximum leeway to support enforcement profiles (below)
and ensures that the existing ability of sanitizers is extended,
and that constant expressions containing bugs become ill-formed.

> ## Enforcement Profiles

Author: This phrase doesn't convey what it is supposed to and can be confused with the strategies. Rename. Suggestion: Dynamic Contract Enforcement Profiles.

{
  fmt::print(stderr, std::forward<args>(parameters)...);
  fmt::print("Try --help\n");
  std::abort();
}
```

Usage couldn't be much simpler:

```c++
  if (actual_num_params != expected_num_params) {
    eg_fatal("Wrong number of arguments provided. Expected={}; Actual={}\n", actual_num_params, expected_num_params);
  }
```

This approach is often discouraged in C++ programs. The main reason is that it typically
bypasses destructors. That's not such a worry if the program is already in a bad
state, so terminating in reaction to contract violations is relatively palatable.
And destructors in a modern, well designed system are only important to a running
process: memory, file descriptors and peripherals should all be freed up by
the system once the owning process is ended.

So with caveats, this can be the best approach for reacting to violations of the
End User Contract -- as well as contracts which cause UB. In profile,
Safety-Critical System With Redundancy, the requirement to 'fail fast' can be well
served by this approach. And in profile, Business-Critical Systems, the problems
associated with bypassing destructors may not be significant.

> ### Push UB Onto The End User

Author: Note that this changes the terms of the End User Contract

### Don't Optimize Until You Sanitize

C++ toolchains increasingly optimise programs assuming that they are correct.
Optimisations are not free. They demand strict adherence to contracts.
Historically, UB has been difficult to detect.

Fortunately, toolchains now also provide facilities for detecting some UB.
These include dynamic analysis tools known as sanitizers
which instrument code to test for bugs.
The beauty of undefined behaviour is that it allows this -- and every other valid
bug-hunting tool -- to be used in conforming code.

Meanwhile, modern software development practices emphasise automated testing regimes.
It is not uncomment for most of the APIs in a program
to be tested for provider C++ API Contract violation frequently during development.
Therefore, there is strong reason to use sanitizers as a matter of good practice.

If you wish to enable compiler optimisations in the program you provide to users
you are strongly advised to test your program code with sanitizers.
Combined with the approaches detained in this document, this will lead to significantly
less defective software without the necessity to compromise on safety or performance.

> ### Don't Rely on Sanitizers Alone

Author: Follow up with a call to tool chain implementers to enable more warnings by default and enable sanitizers by default in -O0. Discuss legacy projects and the harm they do for new projects.
